include/aws/common/resource_name.h

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2010-2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+#pragma once
+
+#include <aws/common/byte_buf.h>
+#include <aws/common/common.h>
+
+struct aws_resource_name {
+    struct aws_byte_cursor partition;
+    struct aws_byte_cursor service;
+    struct aws_byte_cursor region;
+    struct aws_byte_cursor account_id;
+    struct aws_byte_cursor resource_id;
+};
+
+AWS_EXTERN_C_BEGIN
+
+/**
+    Given an ARN "Amazon Resource Name" represented as an in memory a
+    structure representing the parts
+*/
+AWS_COMMON_API
+int aws_resource_name_init_from_cur(struct aws_resource_name *arn, const struct aws_byte_cursor *string);
+
+/**
+    Serializes an ARN structure into the lexical string format
+*/
+AWS_COMMON_API
+int aws_byte_buf_append_resource_name(struct aws_byte_buf *buf, const struct aws_resource_name *arn);
```

Comment:
Should we just have a resource_name -> buf function? Maybe in addition to this one? Especially for like debug printing these or interpolating them into JSON or something.

Reply:
should we have a mode on this API that will give the user a length on how much memory to allocate so they don't have to use a dynamically resizing buffer? I don't like making allocation decisions for users we don't have to make.

Reply:
> Should we just have a resource_name -> buf function? Maybe in addition to this one? Especially for like debug printing these or interpolating them into JSON or something.
I wouldn't be sure where to anchor a resource_name -> byte_buf function quite yet. Without being aware of the the use case details yet we can add this later.

> should we have a mode on this API that will give the user a length on how much memory to allocate so they don't have to use a dynamically resizing buffer? I don't like making allocation decisions for users we don't have to make.
Removed all resizing and allocations from functions.